ls1intum/Artemis
# `Notifications`: Add separate notification and setting for changed programming test cases

### Checklist
#### General
#### Server
#### Client

### Motivation and Context
@krusche requested a new setting to toggle the following notifications and add additional information (course + exercise title) to them
![image](https://user-images.githubusercontent.com/65814168/142234445-f3288290-dd17-4cc5-a219-8f2b5955d328.png)
![image](https://user-images.githubusercontent.com/65814168/142602124-9b901167-cb8f-4fa1-a924-df16c1584710.png)

### Description
The title of the notification is the part that makes it possible to filter them by the notification settings. So I can only add the dynamic information (course + exercise title) to the notification text.

This type of notification was part of the notification type "ExerciseUpdated" so I had to separate it and create a new notification type for it. (To be able to filter it out explicitly)

I.e. I simply extracted and refactored this notification into a new one and added the corresponding settings for instructors.

new Notification:
![image](https://user-images.githubusercontent.com/65814168/142601464-595c043a-585e-4ad2-b1fb-2c128b9704e2.png)


### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise

https://user-images.githubusercontent.com/65814168/142259102-75565ab0-be05-4622-8df6-b2d8b6038c0d.mp4
(the settings look a bit different since I added a new editor exclusive group)
(Simply follow the video)

1) Create a new programming exercise or use an already existing one
2) Participate with the student (i.e. start the exercise, change something and save your changes) (This step is mandatory without it no notification will be created!)
3) Change something in the test classes as an instructor. (My prefered way is to use the code editor -> change the tab to TEST)
4) Verify that a notification is created (after you have changed the test cases) and make sure it looks like in the screenshots.
(5) Toggle the option/setting on and off to check if the corresponding setting works as expected)

### Review Progress
#### Code Review
#### Manual Tests

### Test Coverage
Client tests all pass (locally) 
![image](https://user-images.githubusercontent.com/65814168/142231070-1570b0dd-5d3e-41eb-8d06-73756e978845.png)

I added corresponding tests for the new notification type (e.g. in the GroupNotificationFactoryTest class)

I simply refactored a constant for the notification type for ExerciseUpdates into a new separate type therefore the logic & tests stay the same.
